testsuite/tools/expect_test.ml

```diff
@@ -132,11 +132,17 @@ let split_chunks phrases =
 module Compiler_messages = struct
   let print_loc ppf (loc : Location.t) =
     let startchar = loc.loc_start.pos_cnum - loc.loc_start.pos_bol in
-    let endchar = loc.loc_end.pos_cnum - loc.loc_start.pos_cnum + startchar in
+    let endchar = loc.loc_end.pos_cnum - loc.loc_start.pos_bol in
     Format.fprintf ppf "Line _";
     if startchar >= 0 then
       Format.fprintf ppf ", characters %d-%d" startchar endchar;
-    Format.fprintf ppf ":@,"
+    Format.fprintf ppf ":@.";
+    begin match !Location.input_lexbuf with
+    | None -> ()
+    | Some lexbuf ->
+        Location.show_code_at_location ppf lexbuf loc
+    end;
+    ()
```

Comment:
Wouldn't it be plain simpler to call the existing highlight_dumb function without the ~print_chars flag, which would change the output from:

```
Line _, characters 1-2:
  text
  ^^
```

to:

```
Characters 1-2:
  text
  ^^
```

Reply:
So we tried that first, and the result is something like

```
Characters 346-373:
```

and it is not stable to re-running the script again with `-principal` filled in, so it doesn't actually work.

Reply:
Ok, expect_tests works on the whole file, so positions are from the beginning of the file, whereas the toplevel works phrase by phrase. Fair enough.